File: Difficulty/Hard/[44] Wildcard Matching/solution.cpp
class Solution {
public:
    bool isMatch(string s, string p) {
        int sLen = s.size(), pLen = p.size(), i = 0, j = 0, sTrack = -1, pTrack = -1;
        while (i < sLen){
            if ((s[i] == p[j]) || (p[j] == '?')) {i++; j++; continue;}
            if (p[j] == '*') {sTrack=i; pTrack=j++; continue;}
            if (pTrack >= 0) {j = pTrack+1; i=++sTrack; continue;}
            return false;
        }
        while(p[j] == '*') j++;
        return j==pLen;
    }
};

File: Difficulty/Hard/[44] Wildcard Matching/solution.py
#
# @lc app=leetcode id=44 lang=python3
#
# [44] Wildcard Matching
#
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        sLen, pLen, i, j, sTrack, pTrack = len(s), len(p), 0, 0, -1, -1
        while i < sLen:
            if j < pLen and (s[i] == p[j] or p[j] == '?'):
                i += 1
                j += 1
            elif j < pLen and p[j] == '*':
                sTrack = i
                pTrack = j
                j += 1
            elif pTrack >= 0:
                j = pTrack + 1
                sTrack += 1
                i = sTrack
            else:
                return False
        while j < len(p) and p[j] == '*':
            j += 1
        return j == pLen
